src/runtime/intern.cs

```diff
@@ -0,0 +1,84 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Python.Runtime
+{
+    static partial class InternString
+    {
+        class PyIdentifierDefinition
+        {
+            public string Name { get; private set; }
+
+            public PyIdentifierDefinition(string name)
+            {
+                Name = name;
+            }
+        }
+
+        private static Dictionary<string, IntPtr> _string2interns;
+        private static Dictionary<IntPtr, string> _intern2strings;
+
+        static InternString()
+        {
+            var identifierNames = typeof(PyIdentifier).GetFields().Select(fi => fi.Name);
+            var validNames = new HashSet<string>(identifierNames);
+            if (validNames.Count != _builtinNames.Length)
+            {
+                throw new InvalidOperationException("Identifiers args not matching");
+            }
+            foreach (var name in _builtinNames)
+            {
+                if (!validNames.Contains(name))
+                {
+                    throw new InvalidOperationException($"{name} is not declared");
+                }
+            }
+        }
+
+        public static void Initialize()
+        {
+            _string2interns = new Dictionary<string, IntPtr>();
+            _intern2strings = new Dictionary<IntPtr, string>();
+
+            Type type = typeof(PyIdentifier);
+            foreach (string name in _builtinNames)
+            {
+                IntPtr op = Runtime.PyUnicode_InternFromString(name);
+                SetIntern(name, op);
+                type.GetField(name).SetValue(null, op);
+            }
+        }
+
+        public static void Shutdown()
+        {
+            foreach (var ptr in _intern2strings.Keys)
+            {
+                Runtime.XDecref(ptr);
+            }
+            _string2interns = null;
+            _intern2strings = null;
+        }
+
+        public static string GetManagedString(IntPtr op)
+        {
+            string s;
+            if (TryGetInterned(op, out s))
+            {
+                return s;
+            }
+            return Runtime.GetManagedString(op);
+        }
```

Comment:
Having pointers ready for method names Python.NET uses internally is a good idea. E.g. `PyIdentifier.__class__` we might call very often.

I am not so sure about having a separate bidirectional mapping just for them though. A generic cache for conversions might be good, but not when it just caches a few specific strings.

I think the `GetManagedString`, `SetIntern` and `TryGetInterned` should all be removed.

Reply:
`GetManagedString` is use for 
```c#
string name = InternString.GetManagedString(key);
switch (name)
{
    case "__doc__":
        // ...
}
```

Reply:
@amos402 I meant the new one removed, and the old one kept.

Reply:
You mean replace the old one with the new one?
It's better to keep both. Calling `Runtime.GetManagedString` inside the `Intern.GetManagedString` just a fallback if that's not an interned string.
Like the exmple above, the `key` may not be an intern string.
```python
obj.__doc__ # intern string
getattr(obj, "__doc__") # intern string
getattr(obj, "__doc" + "__") # new string
```
You can use `Runtime.GetManagedString` directly if you certainly know that string is not an interned string so it can skip the hash checking.

Reply:
I mean completely remove this new one. I believe it has very low performance benefit, and may even harm user code performance. And the code is also very complicated, which does not seem to be justified by the potential benefit.

Reply:
On the contrary it improve the performance, you should check the usages.

Reply:
@amos402 the only usage I see is the one in `ModuleObject`, and I don't think accessing module object members is a hot path, that needs special optimization (again, considering the complexity).

Reply:
A path is not hot means it can squander computation randomly? I object to this kind of point resolutely. That's the primary cause of why the master branch has a low performance.
Besides it's not a specialized function, you can use it in other places. It used one place here doesn't mean other branches would use one place too, doesn't mean in the future it would always has one place too.